### PR TITLE
RavenDB-17872 - Get and update compare exchange metadata in the session.

### PR DESCRIPTION
**Pull request by shaharhikri**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17872/Get-and-update-compare-exchange-metadata-in-the-session

### Additional description

Fix - compare-exchange metadata can't be updated in the session.
### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 1]
```diff
@@ -452,6 +470,35 @@ internal bool HasChanged(CompareExchangeValue<BlittableJsonReaderObject> origina
...
+                            newMetadata.TryGetValue(key, out string newVal);
+                            oldMetadata.TryGetValue(key, out string oldVal);
+                            if (oldVal != newVal)
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                return false;
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: What happened if `oldMetadata != null` but `newMetadata == null`?

Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: Fixed it

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 2]
```diff
@@ -452,6 +470,44 @@ internal bool HasChanged(CompareExchangeValue<BlittableJsonReaderObject> origina
...
+                }
+                else
+                {
+                    if (oldMetadataNullOrEmpty) // newMetadataNullOrEmpty == false && oldMetadataNullOrEmpty == true
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        foreach (var key in newMetadata.Keys) // newMetadataNullOrEmpty == false && oldMetadataNullOrEmpty == false
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: what if oldMetadata has more keys than newMetadata?

Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: Done

File: test/SlowTests/Issues/RavenDB-17872.cs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
+            {
+                var entity = await session.Advanced.ClusterTransaction.GetCompareExchangeValueAsync<TestObj>(id);
+                Assert.Equal(1, entity.Metadata.Count);
+                Assert.Contains(metadataPropName + "1", entity.Metadata.Keys);
+                Assert.Equal(metadataValue + "1", entity.Metadata[metadataPropName + "1"]);
+            }
+        }
+
+        private class TestObj
```
Comment (karmeli87) [inline test/SlowTests/Issues/RavenDB-17872.cs:253]: let's add here some nested value and have a test for that as well

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 4]
```diff
@@ -320,16 +321,31 @@ internal CompareExchangeValue<T> GetValue<T>(DocumentConventions conventions)
...
+                                    {
+                                        var key = kvp.Key;
+                                        var val = kvp.Value;
+                                        metadata[key] = val;
+                                    }
+                                }
                             }
 
-                            var value = new CompareExchangeValue<T>(_key, _index, entity);
+                            var value = new CompareExchangeValue<T>(_key, _index, entity, metadata);
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: why not to pass here `originalValue?.Metadata` directly?

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 5]
```diff
@@ -321,16 +322,31 @@ internal CompareExchangeValue<T> GetValue<T>(DocumentConventions conventions)
...
+                                if (_originalValue.Metadata != null)
+                                {
+                                    metadata = new MetadataAsDictionary();
+                                    foreach (var kvp in _originalValue.Metadata)
+                                    {
+                                        var key = kvp.Key;
+                                        var val = kvp.Value;
+                                        metadata[key] = val;
+                                    }
+                                }
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: why do we need that? and not passing `_originalValue?.Metadata` directly?

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 6]
```diff
@@ -391,11 +407,19 @@ internal ICommandData GetCommand(DocumentConventions conventions, JsonOperationC
...
+                        _originalValue?.Value.TryGet(Constants.Documents.Metadata.Key, out metadata);
                         var metadataHasChanged = false;
                         if (_value.HasMetadata && _value.Metadata.Count != 0)
                         {
-                            metadata = PrepareMetadataForPut(_key, _value.Metadata, conventions, context);
-                            metadataHasChanged = InMemoryDocumentSessionOperations.UpdateMetadataModifications(_value?.Metadata, metadata);
+                            if (metadata == null)
+                            {
+                                metadataHasChanged = true;
+                                metadata = PrepareMetadataForPut(_key, _value.Metadata, conventions, context);
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: `PrepareMetadataForPut` should be called for both cases, no?

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: Because you want to add it the modification, not modify it

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 6 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: No, in case it already exist you only need to add the modifications to it

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 5 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: It works also with the ```_originalValue?.Metadata``` instead, but I don't know if it's a correct approach.

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 7]
```diff
@@ -320,6 +322,7 @@ internal CompareExchangeValue<T> GetValue<T>(DocumentConventions conventions)
                                 throw new InvalidOperationException("Value cannot be null.");
 
                             T entity = default;
+                            IMetadataDictionary metadata = null;
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: remove this

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 8]
```diff
@@ -390,19 +393,29 @@ internal ICommandData GetCommand(DocumentConventions conventions, JsonOperationC
...
                         BlittableJsonReaderObject metadata = null;
+                        _originalValue?.Value.TryGet(Constants.Documents.Metadata.Key, out metadata);
+                        var metadataHasChanged = false;
                         if (_value.HasMetadata && _value.Metadata.Count != 0)
                         {
-                            metadata = PrepareMetadataForPut(_key, _value.Metadata, conventions, context);
+                            if (metadata == null)
+                            {
+                                metadataHasChanged = true;
+                                metadata = PrepareMetadataForPut(_key, _value.Metadata, conventions, context); //create new metadata (because there wasn't any metadata before)
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:402]: inside `PrepareMetadataForPut` you have some validation that might be relevant.

Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:402]: Done

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 7 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs]: Done

**Review by karmeli87 — APPROVED** [review]

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs [hunk 9]
```diff
@@ -97,7 +99,7 @@ protected async Task<CompareExchangeValue<T>> GetCompareExchangeValueAsyncIntern
             {
                 _session.IncrementRequestCount();
 
-                var value = await _session.Operations.SendAsync(new GetCompareExchangeValueOperation<BlittableJsonReaderObject>(key, materializeMetadata: false), sessionInfo: _session._sessionInfo, token: token).ConfigureAwait(false);
+                var value = await _session.Operations.SendAsync(new GetCompareExchangeValueOperation<BlittableJsonReaderObject>(key, materializeMetadata: true), sessionInfo: _session._sessionInfo, token: token).ConfigureAwait(false);
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:102]: can you explain why this change is needed?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Client/Documents/Session/ClusterTransactionSession.cs (hunk 9 above)
Comment (shaharhikri) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:102]: Yes,
From what I see now it does not have any influence on the tests,
The command response has metadata which is represented as a 'MetadataAsDictionary'.
If the 'materializeMetadata' is false, the MetadataAsDictionary object gets blittable in its constructor, and every time you try to get a value from the 'MetadataAsDictionary', the 'MetadataAsDictionary' gives you the value from the blittable,
and if it's true, the 'MetadataAsDictionary' object has an actual dictionary and its kvp's initialized (copied) from the blittable.

Comment (karmeli87) [reply src/Raven.Client/Documents/Session/ClusterTransactionSession.cs:102]: so we want to revert this specific change? @ppekrol
